Automated Dihedral Analysis

cadeduckworth opened the pull request:
- Adding automation functionality for dihedral analysis
- Automation function for dihedral analysis includes retrieval of dihedral atom groups and visual representation of data via seaborn-catplot violin plots
- close #236

orbeckst commented on mdpow/analysis/dihedral.py:
```diff
@@ -110,3 +127,257 @@ def _single_frame(self):
             self.results[k] = self._res_dict[k]
+
+# start automation section
+    def directory_paths(parent_directory=None, csv=None):
```
move function to workflows

orbeckst commented on mdpow/analysis/dihedral.py:
```diff
@@ -110,3 +127,257 @@ def _single_frame(self):
+
+        return directory_paths
+
+def directory_iteration(dirpaths, figdir=None, padding=45, width=0.9,
```
move function to workflows

orbeckst reviewed (changes requested):
Move all automation into a separate module: `mdpow/analysis/workflows/base.py`.

Add documentation. Need to include it in the html docs.

Add tests.

orbeckst commented:
This PR requires PR #216 to be merged, otherwise analysis runs twice.

cadeduckworth commented on mdpow/analysis/dihedral.py:
```diff
@@ -110,3 +127,257 @@ def _single_frame(self):
+    g = sns.catplot(data=df, x="lambda", y="dihedral", hue="solvent", col="interaction",
+                    kind="violin", split=True, width=width, inner=None, cut=0,
+                    linewidth=0.5,
+                    hue_order=["water", "octanol"], col_order=["Coulomb", "VDW"],
```
Change was made.

`hue_order=[df.solvent.iloc[0], df.solvent.iloc[-1]]`

cadeduckworth commented on mdpow/analysis/automation/dihedral/automated_dihedral_analysis.py:
```diff
@@ -155,6 +155,49 @@ def dihedral_groups_ensemble(bonds, datadir,
+       as input before padding the angles to optionaly save the raw
+       data before plotting dihedral angle frequencies as KDE violins
+       with dihedral_violins(). Given a parent directory, creates subdirectory
+       for molecule, saves fully sampled csv and per dihedral atom group selection.
```
fix wording in docstring

orbeckst commented:
I'll look at this one again when #218 and #216 are addressed.

orbeckst commented:
For test files: Create a set of data with XTC files that only contain ~10 frames (evenly spaced along the trajectory). Include only TPR files necessary for the test (eg only Coulomb lambda 0).

orbeckst commented:
@cadeduckworth please _resolve conflicts_ now that PR #216 has been merged. This should get this PR up-to-date.

orbeckst reviewed (changes requested):
TODO
- Move all automation into a separate module: mdpow/analysis/workflows ; have a base.py for general stuff and a dihedrals.py for dihedrals specific
- Add documentation.  Show how one uses your code. Use test files as examples so that someone could directly try it out. These howto-docs also give you an idea how to organize and structure.
- Add tests. In particular, create a minimal test set. https://github.com/Becksteinlab/MDPOW/pull/217#issuecomment-1325646698

cadeduckworth commented:
@orbeckst I created and pushed a minimal testing set made with the SAMPL7 GAFF data for SM25. The trajectories were reduced by a factor of 10. It still seems too large, but will function as a base while I develop the tests, and I will remove data that is unnecessary to reduce the size once I am done.

cadeduckworth commented:
The tests are failing because of an error with importing rdkit. Do I need to import rdkit in the testing file as well? 

The error occurs when the test tries to import dihedrals.py for use in the test script:
`from rdkit import Chem`  in dihedrals.py causes the error

cadeduckworth commented:
updated ci.yaml to include pip install of rdkit for use with testing automated dihedral analysis:
`mdpow.analysis.workflows.dihedrals`

cadeduckworth commented:
@orbeckst
For the last test I wrote for using the `directory_iteration` function in base.py (finds the data and then iterates automated dihedral analysis across the directories for an entire project) I realized it might make more sense to restructure the data in testing resources and possibly change the resname in the topologies (they are all 'UNK' in the SAMPL7 datasets). 

Otherwise, the test might become overly complicated from trying to include a 'fix' for the preexisting testing resources and directories to look like results from a typical MDPOW project, and it might not be clear what the test is actually doing.

I included another reduced dataset (SM26) from the SAMPL7/GAFF set to mimic a mini project.
Maybe the `directory_iteration` script itself needs to be more flexible. 
What would you suggest?

orbeckst commented on dir.csv:
```diff
@@ -0,0 +1,6 @@
+molecule,resname,path
```
needs to be somewhere inside the tests, not top level

cadeduckworth replied:
I think this was an accident from running tests, or was somehow saved during testing of the workflows base module and not removed as a temporary file. I will look into it.

orbeckst commented on doc/sphinx/source/analysis.txt:
```diff
@@ -24,6 +24,7 @@ all FEP simulations. They then make data available (typically, as a
 
    analysis/solvation
    analysis/dihedral
+   analysis/hbond
```
remove – no hbond in this PR (I hope)

cadeduckworth replied:
no hbond in this PR, updates to the docs were misplaced and have since been deleted

orbeckst commented on doc/sphinx/source/analysis.txt:
```diff
@@ -48,6 +49,26 @@ in more detail in [Lescoulie2021]_.
+Analysis workflows
+------------------
+
+.. versionadded:: ???
```
next minor release - we are at 0.8.0 so 0.9.0

orbeckst commented on doc/sphinx/source/analysis.txt:
```diff
@@ -48,6 +49,26 @@ in more detail in [Lescoulie2021]_.
+
+   analysis/workflows/base
+   analysis/workflows/dihedrals
+   analysis/workflows/hbonds
```
hbonds?

orbeckst commented on mdpow/analysis/workflows/base.py:
```diff
@@ -0,0 +1,156 @@
+
+#import will need to change based on new naming convention for ada
+#and once the function changes for more general use
+import mdpow.analysis.workflows.dihedrals as ada
```
```python
import .dihedrals as ada
```
when importing from the same package

orbeckst replied:
However, base packages typically don't import other but instead are imported, this avoids circular imports.

Perhaps "base" is not the right name then.

cadeduckworth replied:
Perhaps automation or project automation? Workflows/dihedrals is the automated version of dihedral analysis for one project but workflows/base iterates that automation through a set of projects so it is the top-level automation.

orbeckst commented on mdpow/analysis/workflows/base.py:
```diff
@@ -0,0 +1,156 @@
+    return directory_paths
+
+#needs to be changed for use with other automated workflows
+def directory_iteration(directory_paths, df_save_dir=None, figdir=None, padding=45, width=0.9,
```
This should be made more general so that you can specify which analysis to take. Reduce the code to the generic bits.

cadeduckworth replied:
This will require adaptation of each analysis for use with iteration and desired outcomes (plotting and storing data), which might require a separate workflows module for each (the automation class I made does not make sense because the analyses are not similar enough for significant overlap, it would only confuse those who want to develop their own analyses from the base class). I will start new pull requests to make these changes for each once this pull request is merged and I have more of an idea of how to generalize this function for use with all of the automated physical properties analyses, including those that users might develop on their own.

orbeckst reviewed (changes requested):
Sorry partial review.

See comments. For trajectories: do NOT include any of the dot-files: offsets and locks

orbeckst commented on .github/workflows/ci.yaml:
```diff
@@ -76,6 +76,10 @@ jobs:
     - name: Install pybol (fixture staging tool)
       run: |
          python -m pip install pybol
+         
+    - name: Install rdkit (used in workflows)
+      run: |
+         python -m pip install rdkit
```
add to devtools/conda-envs/test_env.yaml so that it is conda installed instead

orbeckst commented:
It’s also contains file names that are specific to your local installation. If it was needed for tests the. It’s should contain relative paths. But if not needed then remove!!Am 1/6/23 um 23:27 schrieb Cade Duckworth ***@***.***>:﻿
@cadeduckworth commented on this pull request.



In dir.csv:
> @@ -0,0 +1,6 @@
+molecule,resname,path

I think this was an accident from running tests, or was somehow saved during testing of the workflows base module and not removed as a temporary file. I will look into it.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

cadeduckworth commented:
@orbeckst 
After creating a new environment to investigate the issue with Python=3.7, it seems that pybol might not be supported in 3.7. This is not specified in the error being raised for the test I am trying to debug, but no test will run without error when using pytest locally with this version of Python.

orbeckst commented:
But in other PRs, the 3.7 tests with pybol run. What's the actual error?

orbeckst commented:
Looking at https://github.com/ianmkenney/PyBOL there's nothing that restricts pybol. If anything, the danger is that it doesn't work with _newer_ versions of Python.

cadeduckworth commented:
@orbeckst 
My apologies for the confusion concerning pybol, that was not the issue, but an err on my behalf. After many tests with different versions of software packages in a new environment, I believe I have found the issue, which might also be relevant for other testing issues involving Python 3.7 (for those we specified no version compatibility with Python < 3.8 due to what we thought was a scipy error).

The best explanation I have found is the pytest version = 7.2.0, build = py37h89c1867_0, which only exists in the CI micromamba environment for Python =3.7. Installing that specific pytest build in my local environment (Python=3.7) results in failing and skipped tests, when no other do, similar to the CI.

Starting with (CI) Python 3.8, pytest = 7.2.0 upgrades to build = pyhd8ed1ab_2 (same as my (and CI) Python = 3.10 environment where pytest passes locally). With no other differences in results found from testing other software package versions in my notebook (mdpow, mdanalysis, rdkit, python), this leads me to believe there is some change between builds in the way pytest handles fixtures, based on the way the temp data is generated as written in my tests, when being compared to the data provided for comparison to expected results.

In short, previous versions of the aforementioned packages (mdpow, mdanalysis, rdkit, python) provide the same results as expected in my notebook. However, whenever pytest generates the data using a fixture, the testing data is entirely different using those package versions (but with pytest version = 7.2.0, build = py37h89c1867_0).

orbeckst commented:
Can you check if pytest < 7.2 works? We could blacklist pytest == 7.2.0.

cadeduckworth commented:
@orbeckst 
I will check that, but the current build that works for tests with CI Python>=3.7 is still pytest 7.2. It's only the one build for pytest 7.2, python 3.7 that is not working on my end.

cadeduckworth commented:
@orbeckst 
Available version that is pytest < 7.2.0=py37h89c1867_0 (build we have issues with) is pytest=6.2.5-py37h89c1867_3. This one causes the same problem. Pytest=7.2.0-pyhd8ed1ab_2 (current passing version for local and CI) can only run on Python<=3.8. I can check in their release information for the exact changes and write the tests in a way that circumvents the unsupported issue if we will continue to support python 3.7

orbeckst commented:
If this the case then let’s add an xfail if the Python version is 3.7. See the docs for pytest for how to decorate tests with xfail or skipif.

orbeckst commented:
Will probably be pytest.mark.skipif or similar.

cadeduckworth commented:
@orbeckst 
I think it will be:
`@pytest.mark.skipif(sys.version_info < (3, 8), reason="pytest##, build## gives wrong answers")`

I will add this and then move everything pertaining to workflows/base.py (directory paths and iteration module) to a separate PR because it now lives in Automated Dihedral Analysis (PR#217) and Automated Solvation Shell Analysis (PR#227). Once cleanly removed, this pull request should be ready for review!

cadeduckworth commented on devtools/conda-envs/test_env.yaml:
```diff
@@ -15,9 +15,10 @@ dependencies:
 - numkit
 - gromacswrapper
 - alchemlyb <2
+- rdkit
```
add to PR#229 workflows-base

cadeduckworth commented on .gitignore:
```diff
@@ -12,3 +12,6 @@ dist
 doc/examples/benzene/Equilibrium/
 .coverage
 coverage.xml
+*.lock
```
add to PR#229 workflows-base

cadeduckworth commented on doc/requirements.txt:
```diff
@@ -6,3 +6,6 @@ pyyaml
 GromacsWrapper>=0.5.1
 alchemlyb
 mdanalysis
+rdkit
```
add to PR#229 workflows-base

cadeduckworth commented on doc/sphinx/source/analysis.txt:
```diff
@@ -20,12 +20,11 @@ all FEP simulations. They then make data available (typically, as a
 
 .. toctree::
-   :maxdepth: 1
+   :maxdepth: 2
```
revert changes

cadeduckworth commented on mdpow/tests/testing_resources/manifest.yml:
```diff
@@ -45,3 +45,9 @@ example_FEP:
         - full_transfer
     files:
         - []
+
```
this might need to change after having moved test data

orbeckst reviewed (changes requested):
- The XTC files are still quite big in size with 10 MB per trajectory. How many frames does each XTC contain? Can we shrink them by another factor of 10?
- workflows/dihedrals imports seaborn so this also needs to be added to the requirements/env yml/setup.py
- all new package requirements should also be added to `setup.py`
- `workflows/dihedrals.py` (comments here because GitHub refuses to let me see the file "Oops, something went wrong. Retry.")
  - merge `add_hydrogens()` code into `rdkit_conversion()`
  - `dihedral_indices()` docs: format the part on *SMARTS* more nicely, see https://mdpow--217.org.readthedocs.build/en/217/workflows/dihedrals.html#mdpow.workflows.dihedrals.dihedral_indices ; perhaps a description list or bulleted list where each SMARTS component is shown as code, 
    - `[!#1]` : any atom, not Hydrogen 
    - `~`  : any bond
    - ... etc
  - If `dihedral_indices` and `dihedral_groups` use the same default SMARTS string then it makes sense to define it at the module level as
    ```python
    #: Default SMARTS string to define relevant dihedral atom groups.
    #: - `!#1]` : any atom, not Hydrogen
    #: - `~`  : any bond
    #: ...
    SMARTS_DEFAULT = '[!#1]~[!$(*#*)&!D1]-!@[!$(*#*)&!D1]~[!#1]'
    ```
    Then you can use `SMARTS_DEFAULT` as the `SMARTS=SMARTS_DEFAULT` in the kwarg and in the docs you add 
    ```reST
    :autodata:`SMARTS_DEFAULT`
    ```
    at the top and in the function docs you can just link to 
    ```reST
    The default SMARTS string is described in detail under :data:`SMARTS_DEFAULT`.
    ```
    In this way you keep your code [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).
  - For all your function documentation, *state what the function returns*: that's very important for anyone wanting to use the functions. I think you can simply add a `:returns:` block to the docs.
  - `save_df()` should return the filename of the file that it saved to (also test return value in tests). Without know where files are saved, it gets very difficult to manage them in bigger workflows.
  - `periodic_angle()`
    -  the `padding=45` argument should take a float, makes no sense to restrict it to int (possibly just fix the docs).
    - The Example should be code-formatted https://mdpow--217.org.readthedocs.build/en/217/workflows/dihedrals.html#mdpow.workflows.dihedrals.periodic_angle
  - `dihedral_violins()`:
    - Separate the docs about the return value from the "subject" = "Plots distributions of dihedral angles for one dihedral atom group       as violin plots". Generally speaking, we like to write docs as 
      - first line = shot synopsis, 
      - empty line, 
      - paragraph with description, 
      - lists for args and 
      - kwargs, 
      - return values
      - examples
      - other notes
  - `plot_violins()`: break up doc string as described above, define arguments in docs, especially as this is a high-level function a user might call. You can use [`.. SeeAlso::`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-seealso) directives to link to other related functions, too.
  - `automated_dihedral_analysis()`
    - break up docs
    - `*dataframe : Pandas DataFrame*` --> `*dataframe* : :class:'pandas DataFrame'` (only I had to use here apostrophe instead of the correct backtic because GitHub markdown does not let me write it out) 
  - format example as code https://mdpow--217.org.readthedocs.build/en/217/workflows/dihedrals.html#mdpow.workflows.dihedrals.automated_dihedral_analysis
  - In the reST documentation, put the docs for `automated_dihedral_analysis()` *first* as this is what the users will first want to look at. You can even consider making subsections to separate the user code from the components. Think what would be most helpful for a new user.

orbeckst reviewed (changes requested):
- The XTC files are still quite big in size with 10 MB per trajectory. How many frames does each XTC contain? Can we shrink them by another factor of 10?
- workflows/dihedrals imports seaborn so this also needs to be added to the requirements/env yml/setup.py
- all new package requirements should also be added to `setup.py`
- `workflows/dihedrals.py` (comments here because GitHub refuses to let me see the file "Oops, something went wrong. Retry.")
  - merge `add_hydrogens()` code into `rdkit_conversion()`
  - `dihedral_indices()` docs: format the part on *SMARTS* more nicely, see https://mdpow--217.org.readthedocs.build/en/217/workflows/dihedrals.html#mdpow.workflows.dihedrals.dihedral_indices ; perhaps a description list or bulleted list where each SMARTS component is shown as code, 
    - `[!#1]` : any atom, not Hydrogen 
    - `~`  : any bond
    - ... etc
  - If `dihedral_indices` and `dihedral_groups` use the same default SMARTS string then it makes sense to define it at the module level as
    ```python
    #: Default SMARTS string to define relevant dihedral atom groups.
    #: - `!#1]` : any atom, not Hydrogen
    #: - `~`  : any bond
    #: ...
    SMARTS_DEFAULT = '[!#1]~[!$(*#*)&!D1]-!@[!$(*#*)&!D1]~[!#1]'
    ```
    Then you can use `SMARTS_DEFAULT` as the `SMARTS=SMARTS_DEFAULT` in the kwarg and in the docs you add 
    ```reST
    :autodata:`SMARTS_DEFAULT`
    ```
    at the top and in the function docs you can just link to 
    ```reST
    The default SMARTS string is described in detail under :data:`SMARTS_DEFAULT`.
    ```
    In this way you keep your code [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).
  - For all your function documentation, *state what the function returns*: that's very important for anyone wanting to use the functions. I think you can simply add a `:returns:` block to the docs.
  - `save_df()` should return the filename of the file that it saved to (also test return value in tests). Without know where files are saved, it gets very difficult to manage them in bigger workflows.
  - `periodic_angle()`
    -  the `padding=45` argument should take a float, makes no sense to restrict it to int (possibly just fix the docs).
    - The Example should be code-formatted https://mdpow--217.org.readthedocs.build/en/217/workflows/dihedrals.html#mdpow.workflows.dihedrals.periodic_angle
  - `dihedral_violins()`:
    - Separate the docs about the return value from the "subject" = "Plots distributions of dihedral angles for one dihedral atom group       as violin plots". Generally speaking, we like to write docs as 
      - first line = shot synopsis, 
      - empty line, 
      - paragraph with description, 
      - lists for args and 
      - kwargs, 
      - return values
      - examples
      - other notes
  - `plot_violins()`: break up doc string as described above, define arguments in docs, especially as this is a high-level function a user might call. You can use [`.. SeeAlso::`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-seealso) directives to link to other related functions, too.
  - `automated_dihedral_analysis()`
    - break up docs
    - `*dataframe : Pandas DataFrame*` --> `*dataframe* : :class:'pandas DataFrame'` (only I had to use here apostrophe instead of the correct backtic because GitHub markdown does not let me write it out) 
  - format example as code https://mdpow--217.org.readthedocs.build/en/217/workflows/dihedrals.html#mdpow.workflows.dihedrals.automated_dihedral_analysis
  - In the reST documentation, put the docs for `automated_dihedral_analysis()` *first* as this is what the users will first want to look at. You can even consider making subsections to separate the user code from the components. Think what would be most helpful for a new user.

cadeduckworth commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -0,0 +1,188 @@
+            assert groups[i].all() == self.check_groups[i].all()
+            i+=1
+
+    @pytest.mark.skipif(sys.version_info < (3, 8), reason="scipy circvar gives wrong answers")
```
This was one of the original testing issues we fixed for DihedralAnalysis, from which I copied over the `skipif` because `scipy.stats` is used in the same way here (caused the same failure). I suggested in an earlier comment that it might be pytest because of the similarity with it only being Python < 3.8. I will have to set up an environment to investigate and figure out if it is pytest or scipy.

orbeckst replied:
Any clearer about this? At minimum. add a comment that states what we know/suspect.

orbeckst replied:
Add issue number, add more text to be specific ... whoeever reads this has no clue what "nature of this problem" means.

Read the comments through the eyes of someone who is not an insider.

cadeduckworth commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -0,0 +1,188 @@
+        atom_group_indices = dihedrals.dihedral_indices(dirname=SM25_tmp_dir, resname=self.resname)
+        # test user input of SMARTS string
+        atom_group_indices_alt = dihedrals.dihedral_indices(dirname=SM25_tmp_dir, resname=self.resname,
+                                                            SMARTS='[!$(*#*)&!D1]-!@[!$(*#*)&!D1]')
```
All functions that depend on `dihedral_indices` function are by default done with our SMARTS string that returns four-tuples of dihedral atom groups. I will add comments to address these differences in the data, fixtures, and tests.

This alternative SMARTS string is the original provided by Dr. Iorga that I built upon to extend from two-tuples to four-tuples; this one returns the two-tuples. I believe it is a good example for testing the MDA/RDKit conversion functionality because we are expecting that the 1-2 indices of the four-tuples should be the same as the two-tuples. Also, I can add an `assert` line or a separate test to verify this and really make sure that MDA-RDKit-SMARTS is doing what we want it to all the time.

cadeduckworth commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -0,0 +1,188 @@
+RESOURCES = pathlib.PurePath(resource_filename(__name__, 'testing_resources'))
+MANIFEST = RESOURCES / "manifest.yml"
+
+@pytest.fixture(scope="function")
```
In tests at [line 166 and line 184](https://github.com/Becksteinlab/MDPOW/pull/217/files/e8a7c650d20a515a4bbb3c70c73ea7bfbc17c36b#diff-56647fb1069337c6213bb0005ef40522d16da8058827420c9e87e6f870616c58), the tests are using `SM25_tmp_dir` created by the function-scoped fixture at line 40 as the directory to save compressed .csv.bz2 files or .pdf plot figures. I will read more in the pytest docs about the best way to scope these fixtures and raise an issue.

Generally, the other existing test modules still need an update for things we discussed previously, so I can start an mdpow/tests PR and a cleanup PR.

cadeduckworth commented:
after reducing testing set by another factor of 10 (total reduction by factor of 1000), data for comparison in test module will need to be recalculated

cadeduckworth commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -0,0 +1,188 @@
+        solute_names = solute.atoms.names
+        assert solute_names.all() == self.universe_solute_atom_names.all()
+
+    def test_add_hydrogens(self, SM25_tmp_dir):
```
I believe the `add_hydrogens` function was written very early on in development (several months ago) while working on the SMARTS string, and this investigation has served as an important lesson and additional reminder to take better notes!

cadeduckworth commented:
link issue #232

orbeckst commented:
Let me know when the PR is ready for a final(?) review, @cadeduckworth . Good progress!!

cadeduckworth commented:
@orbeckst 
Unless additional tests are necessary (apart from the one requested to test the 'save location return value from save_df'), then I will likely be able to confidently request a final review today.

I do need to test that all input alternatives work, and cover them with tests (i.e., input results as DF or results as csv, running from top-level for downstream padding and plotting) - *reminder for myself*

I have added in-line comments as reminders to split up big 'doc-blocks' and format them per your review requests. There are a few other commented reminders for minor issues that might not necessarily have to be handled in this PR since they are covered in upcoming developments. That may be a discussion better held in a meeting.

I have not looked back into the documentation for the tutorials notebook we added, but that also might be better suited in a separate issue or PR. (issue #232)

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -0,0 +1,215 @@
+        return dirname
+
+    @pytest.fixture(scope="function")
+    def gen_data(self, SM25_tmp_dir):
```
Can you split gen_data into two fixtures
- `dihedral_data`: produces `atom_group_indices, df, df_aug` (runs analysis) and is used when we want to look at the analysis output
- `atom_indices`: produces `atom_group_indices, atom_group_indices_alt` (only does the initial dihedral_indices) and is used when checking indices

cadeduckworth replied:
Responding to question in comment above, I think all workflows related testing data in this PR is 14MB in total.

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -0,0 +1,215 @@
+        assert SM25_tmp_dir / 'SM25' / 'SM25_full_df.bz2'
+                                        
+
+    @pytest.mark.skipif(sys.version_info < (3, 8), reason="scipy circvar gives wrong answers") 
```
comment as above : pytest or scipy?

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -0,0 +1,215 @@
+                                              solvents=('water',))
+        assert SM25_tmp_dir / 'SM25' / 'SM25_C10-C5-S4-O11_violins.pdf'
+
+    def test_DataFrame_input(self, SM25_tmp_dir):
+        test_df = pd.DataFrame([['C1-C2-C3-C4', 'water', 'Coulomb', 0, 0, 60.0],
+                                ['C1-C2-C3-C5', 'water', 'Coulomb', 0, 0, 60.0]],
+                                [1,2],['selection', 'solvent', 'interaction', 'lambda', 'time', 'dihedral'])
+        dihedrals.automated_dihedral_analysis(dirname=SM25_tmp_dir, figdir=SM25_tmp_dir,
+                                              resname=self.resname,
+                                              solvents=('water',), dataframe=test_df)
+
+        with open('mdpow.log', 'r') as file:
+            log_content = file.read()
+            return_value = log_content.find('Proceeding with results DataFrame provided.')
+
+        assert return_value != -1
```
What is tested here?

Looking in the log file is a very indirect way to check for things (and may fail when multiple tests write to it). Is there a more direct to test what you want to test? Doesn't `dihedrals.automated_dihedral_analysis` return something that we can work with?

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -0,0 +1,215 @@
+        assert return_value != -1
+
+    def test_save_results_warning(self, SM25_tmp_dir):
+        dihedrals.automated_dihedral_analysis(dirname=SM25_tmp_dir, df_save_dir=None,
```
same as above: instead of log file searching, have the code raise a warning and then use [pytest.warns](https://docs.pytest.org/en/7.1.x/how-to/capture-warnings.html#warns)
```python
with pytest.warns(UserWarning):
    xxx = dihedrals.automated_dihedral_analysis(dirname=SM25_tmp_dir, df_save_dir=None,

# assert something about xxx, too
```
You can also have `pytest.warns` check the warning message.

If you emit another warning than UserWarning, check for that instead.

cadeduckworth replied:
They are not actually warnings of the Warning type subclass of Exceptions, they are logging.WARNING and logging.INFO messages, that are output or returned to display informative messages to the user and to also be recorded in the logs in case someone forgot to specify a results data frame saving directory and wants to know why their results were not saved.

I changed the testing methods, and they now use pytest's fixture `caplog` because `pytest.warn` only works with objects in the Warning class.

orbeckst reviewed (changes requested):
Almost there, just a few issues with the testing, as far as I can tell.

What is the size of the bundled new files?

cadeduckworth commented:
Some issues need to be unlinked and linked to separate or new PR's after this one is merged
- only issue #236 should be closed with this PR

cadeduckworth commented:
@orbeckst
I will double check docs, tests, and modules, and update if I notice anything, but it appears that everything should be ready for final review.

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -96,17 +101,33 @@ def gen_data(self, SM25_tmp_dir):
     # pre-padding dihedral group statistics
-    DG_O1C2N3S4_mean = -0.13089566578887002
-    DG_O1C2N3S4_var = 0.03473127346296412
+    #DG_O1C2N3S4_mean = -0.13089566578887002
+    #DG_O1C2N3S4_var = 0.03473127346296412
```
What do the commented out values mean? 

With version control we tend to avoid leaving commented out code in there. It's just ambiguous: either there's a good reason to replace it, then it is replaced. Or not... and then one has to work harder.

What's the justification for keeping old and new visible?

cadeduckworth replied:
- old data from most recently reduced test data set preceding the current, forgot to remove after generating new data
- handled

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -23,11 +24,8 @@
-
-# Are any additional tests needed?
-# Top level?
+# NEW TESTING DATA NEEDS TO BE GENERATED FOR circstats
```
I don't understand the comment.

If there's a TODO, raise an issue and reference the issue number if necessary. Otherwise remove comments. They will just be confusing later.

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -96,17 +101,33 @@ def gen_data(self, SM25_tmp_dir):
+    #DG_C13141520_mean = 90.04076626959608
+    #DG_C13141520_var = 0.8814931303534448
+
+    # new 2-6-23
```
Don't put new/old dates etc in here. It's not going to be helpful outside the narrow context of this PR.

The code needs to stand on its own.

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -151,16 +173,22 @@ def test_dihedral_groups_ensemble(self, gen_data):
+        caplog.clear()
+        caplog.set_level(logging.INFO, logger='mdpow.workflows.dihedrals')
+        dihedrals.save_df(df=dihedral_data[0], df_save_dir=SM25_tmp_dir, molname='SM25')
+        assert f'Results DataFrame saved as {SM25_tmp_dir}/SM25/SM25_full_df.csv.bz2', 'Save location not logged or returned'
```
This assertion is always true — it's an non-empty string.

assert something about caplog

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -151,16 +173,22 @@ def test_dihedral_groups_ensemble(self, gen_data):
+        dihedrals.save_df(df=dihedral_data[0], df_save_dir=SM25_tmp_dir, molname='SM25')
+        assert f'Results DataFrame saved as {SM25_tmp_dir}/SM25/SM25_full_df.csv.bz2', 'Save location not logged or returned'
+
+    # separate issue raised to address nature of this problem
```
more specific / less mysterious comment, please

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -170,46 +198,40 @@ def test_periodic_angle(self, gen_data):
         aug_dh2_mean == pytest.approx(self.ADG_C13141520_mean)
         aug_dh2_var == pytest.approx(self.ADG_C13141520_var)
 
+    # related to errors above, naming convention relies on correct atom indices
```
which errors? more specific

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -170,46 +198,40 @@ def test_periodic_angle(self, gen_data):
-        assert SM25_tmp_dir / 'SM25' / 'SM25_C10-C5-S4-O11_violins.pdf'
+        assert (SM25_tmp_dir / 'SM25' / 'SM25_C10-C5-S4-O11_violins.pdf').exists(), 'PDF file not generated'
+
+    # related to errors above, naming convention relies on correct atom indices
```
which errors? more specific

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -170,46 +198,40 @@ def test_periodic_angle(self, gen_data):
+        plot = dihedrals.automated_dihedral_analysis(dirname=SM25_tmp_dir, figdir=SM25_tmp_dir,
+                                                     resname=self.resname,
+                                                     solvents=('water',), dataframe=test_df)
+        assert plot # get exact object ID
```
Weak assertion. At least something like
```python
assert isinstance(plot, matplotlib.XXXXX), "Failed to generate a matplotlib plot"
```
where XXXX is the correct class.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -323,30 +322,40 @@ def dihedral_groups_ensemble(dirname, atom_group_indices,
-       *df*
-           results :class:`pandas.DataFrame` from
-           :class:`~mdpow.analysis.dihedral.DihedralAnalysis`
+        *df*
```
wrong indentation I think

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -323,30 +322,40 @@ def dihedral_groups_ensemble(dirname, atom_group_indices,
-           Saves the results :class:`pandas.DataFrame` as a compressed
-           csv file (bz2).
+           saves the results :class:`pandas.DataFrame` as a compressed
+           csv file, default: .bz2
```
The function does not return anything, therefore, remove the `:returns:`. We do not have a formal item to describe products of functions.

Instead, describe in the text body above what the output is, just using the sentence here.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -357,27 +366,24 @@ def save_df(df, df_save_dir=None, resname=None, molname=None):
+                 '{newdir}/{molname}_full_df.csv.bz2')
 
-    return
+    #return
```
remove — avoid littering with commented-out code

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -503,18 +512,48 @@ def dihedral_violins(df, width=0.9, solvents=solvents_default):
+       `figdir` using `resname` or `molname` provided and saves violin plot
+       figur for each dihedral atom group separately.
+
+       .. seealso: :func:`~mdpow.workflows.dihedrals.automated_dihedral_analysis`,
+                   :func:`~mdpow.workflows.dihedrals.dihedral_violins`
```
Probably should be rewritten as
```reST
.. seealso: 
   :func:`~mdpow.workflows.dihedrals.automated_dihedral_analysis`,
   :func:`~mdpow.workflows.dihedrals.dihedral_violins`
```
... unless it looks fine in the docs!

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -573,43 +615,52 @@ def automated_dihedral_analysis(dirname=None, df_save_dir=None, figdir=None,
            value in degrees
            default: 45
-           see :func:`~mdpow.workflows.dihedrals.periodic_angle`
+       .. seealso:: :func:`~mdpow.workflows.dihedrals.periodic_angle`
```
indent to be in the same block as the rest of the padding text, otherwise breaks flow

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -573,43 +615,52 @@ def automated_dihedral_analysis(dirname=None, df_save_dir=None, figdir=None,
            width of the violin element (>1 overlaps)
            default: 0.9
-           see :func:`~mdpow.workflows.dihedrals.dihedral_violins`
+       .. seealso:: :func:`~mdpow.workflows.dihedrals.dihedral_violins`
```
indent!

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -573,43 +615,52 @@ def automated_dihedral_analysis(dirname=None, df_save_dir=None, figdir=None,
+           The default interactions are documented under :data:`INTERACTIONS_DEFAULT`
+
+       *start, stop, step*
+           .. seealso:: :class:`~mdpow.analysis.ensemble.EnsembleAnalysis`
```
add minimal text above see also

orbeckst reviewed (changes requested):
Good changes. I found one incorrect assertion, some formatting issues, and please clean up comments/commented out code and make some of the comments more specific so that people after you (or your future self) has a chance understanding what the problem is.

cadeduckworth commented:
@orbeckst All requested changes have been addressed, and relevant or meaningful inline comments that do not fit the scope of this pull request have been moved to new or existing issues.

As a reminder (because I tried, but could not figure out how to do it), issues 232 and 235 need to be unlinked to make sure they are not closed with this PR. Only issue 236 should be closed with this PR.

orbeckst reviewed (approved):
excellent work, good to go 🎉 

I'll squash-merge.

orbeckst commented:
oops, need entry in CHANGES ... will do that before merge

orbeckst commented:
Congratulations @cadeduckworth , your big PR is in MDPOW! 🎉 🍾
